src/tmpfiles/tmpfiles.c

```diff
@@ -2638,13 +2638,22 @@ static int parse_line(const char *fname, unsigned line, const char *buffer, bool
 
         case COPY_FILES:
                 if (!i.argument) {
-                        i.argument = strappend("/usr/share/factory/", i.path);
+                        i.argument = path_join(arg_root, "/usr/share/factory/", i.path);
                         if (!i.argument)
                                 return log_oom();
+
                 } else if (!path_is_absolute(i.argument)) {
                         *invalid_config = true;
                         log_error("[%s:%u] Source path is not absolute.", fname, line);
                         return -EBADMSG;
+
+                } else if (arg_root) {
+                        char *p;
+
+                        p = prefix_root(arg_root, i.argument);
+                        if (!p)
+                                return log_oom();
+                        free_and_replace(i.argument, p);
                 }
```

Comment:
this sounds OK for "C" lines. But I do wonder: this creates an interesting assymetry between the "L" and the "C" lines: both in many ways have similar effects: after the operation ran the specified paths will access the data in the parameter field. But with your change the "L" lines will be taken relative to the host's "/", while the "C" lines are relative to root. Now I am not sure this assymetry matters, but it does make me wonder. Maybe a middle ground would be OK: i.e. never honour the --root= param for neither "L" nor "C" sources, but warn about this loudly: whenever an absolute parameter is specified in combination with --root= we complain but go on, and suggest people rwrite these files with relative source specifications.

Does that make sense? Opinions?

Reply:
Hmm, are you sure? This patch only touches code under `case COPY_FILES`, so it shouldn't change anything for `L`. Am I missing some side effect? And I don't think we can meaningfully say that `L` paths are taken relative to anything. We simply create a link with the specified destination, and any interpretation of the link is only done later.

```console
$ echo 'L+ /etc/os-release2 - - - - /usr/lib/os-release2' | sudo build/systemd-tmpfiles --root=/var/lib/machines/rawhide - --create
$ sudo ls -l /var/lib/machines/rawhide/etc/os-release2
lrwxrwxrwx. 1 root root 20 Jun 19 15:15 /var/lib/machines/rawhide/etc/os-release2 -> /usr/lib/os-release2
```

Reply:
but that's what I meant: right now C and L kind behave the same when it comes to absolute parameter paths: the result you get if you run things with --root= is a symlink or a copy based on the host root, not the value of --root=... with your change this changes: L continues to be relative the host, and C is relative to the image root, if you follow what i mean?

Reply:
> L continues to be relative the host, and C is relative to the image root, if you follow what i mean?

For me, paths created with `--root=` are for *consumption* by something also living in the image root. And for anything which reads the symlink from within image root, `L` is relative to the image root. (We know that absolute symlinks in a chroot are "broken" and direct evaluation without taking the alternative root into account will give a different result. This is a special case of this.)